[FLINK-14017][python] Support to start up Python worker in process mode.

WeiZhong94 opened the pull request:
## What is the purpose of the change

*This pull request reimplements the apache-beam container script `boot.go` in python with some minor changes to support the beam python worker running in process mode and avoid introducing go-lang dependencies to Flink.*

## Brief change log

  - *Add `boot.py`,  a python implementation of `boot.go` in "beam-sdks-python-container" project of Apache Beam for launching the python worker.*
  - *Different from `boot.go` of Apache Beam, `boot.py` supports specify the path of python executable file, and the default pip command is `python -m pip` instead of `/usr/local/bin/pip`.*
  - *Different from `boot.go` of Apache Beam, `boot.py` does not restrict the platform of user-specified apache beam sdk package to `linux`.*

## Verifying this change

This change added tests and can be verified by running `flink/flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (PythonDocs)

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+                f.write(artifact_chunk.data)
+            hash_value = sha256obj.hexdigest()
+        if hash_value != sha256:
+            raise Exception("sha256 not consist!")
```
print the filepath, actual hash, expected hash in the exception message.

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+
+    If the environment variable "pip" exists, using its value.
+
+    :return: The pip executable file location. Default value is "/usr/local/bin/pip".
```
Default value is "python -m pip"?

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+        return "python"
+
+
+def pip_command():
```
pip_location?

WeiZhong94 replied:
In some case it returns a command (e.g. python -m pip) instead of a pip location, maybe "pip_command" makes more sense here?

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+            if force:
+                args = ["install", "--upgrade", "--force-reinstall", "--no-deps",
+                        os.path.join(dir, package)]
+                args[0:0] = pip
```
refactor the code to remove the code duplication.

WeiZhong94 replied:
Removed.

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+        print(e)
+
+# try to install the user-provided dataflow python sdk
+sdk_src_file = "dataflow_python_sdk.tar"
```
dataflow package is not needed and should not be installed

WeiZhong94 replied:
Removed.

dianfu commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,352 @@
+install_extra_packages(files, "extra_packages.txt", staged_dir)
+
+# install the packages created from user-provided setup.py
+pip_install_package(files, staged_dir, "workflow.tar.gz", False, True, None)
```
remove this line?

WeiZhong94 replied:
Removed.

dianfu commented on flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py:
```diff
@@ -0,0 +1,129 @@
+            server.start()
+            return server, port
+
+        artifact_server, artifact_port = start_test_artifact_server()
```
move the start the server logic to setup method?

dianfu commented on flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py:
```diff
@@ -0,0 +1,129 @@
+
+        env = dict(os.environ)
+        env["python"] = sys.executable
+        env["FLINK_BOOT_TESTING"] = "1"
```
move to setup method

WeiZhong94 commented:
@dianfu Thanks for your review! I have addressed your comments and remove the code of dependency management in boot.py because it is hard to use in flink-python. Maybe we should make a new design for flink python dependency management?

dianfu commented:
@WeiZhong94 Thanks a lot for the update. LGTM. +1 from my side.

hequn8128 commented on flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py:
```diff
@@ -0,0 +1,140 @@
+
+    def tearDown(self):
+        self.artifact_server.stop(0)
+        self.provision_server.stop(0)
```
Remove the tmp_dir(semi_persist_dir) here .

WeiZhong94 commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -0,0 +1,155 @@
+        if hash_value != sha256:
+            raise Exception("the sha256 hash value: %s of the downloaded file: %s is not the same"
+                            " as the expected hash value: %s" % (hash_value, file_path, sha256))
+        os.chmod(file_path, int(str(permissions), 8))
```
`int(permissions, 8)` won't get the right value we want. The `permissions` is a int value like `444`(decimal, 4 * 100 + 4 * 10 + 4) and what we want is `0o444`(octal, 4 * 8 * 8 +4 * 8 + 4). So we should transform the int value to string and then parse the string as a octal int value.

WeiZhong94 commented on flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py:
```diff
@@ -0,0 +1,140 @@
+        self.provision_server, self.provision_port = start_test_provision_server()
+
+        self.env = dict(os.environ)
+        self.env["python"] = sys.executable
```
The value of `sys.executable` in `boot.py` comes from here in this test case so we'd better not remove this.

WeiZhong94 commented on flink-python/pyflink/fn_execution/tests/test_process_mode_boot.py:
```diff
@@ -0,0 +1,140 @@
+                return manifest_response
+
+            def GetArtifact(self, request, context):
+                yield artifact_chunks[request.name]
```
The return type of `GetArtifact` has been marked as "stream" in grpc, so this method must return a iterable object. Maybe "yield" is better than "return" here?

WeiZhong94 commented:
@hequn8128 Thanks a lot for you review! I have addressed some comments you left and reply other comments that I think we'd better keep current design.

hequn8128 commented:
@WeiZhong94 Thanks a lot for the update. LGTM.
merging...
